pkg/apis/pipeline/v1alpha1/verificationpolicy_validation.go

```diff
@@ -0,0 +1,58 @@
+/*
+Copyright 2022 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+
+	"github.com/tektoncd/pipeline/pkg/apis/validate"
+	"knative.dev/pkg/apis"
+)
+
+var _ apis.Validatable = (*VerificationPolicy)(nil)
+
+// Validate VerificationPolicy
+func (v *VerificationPolicy) Validate(ctx context.Context) (errs *apis.FieldError) {
+	errs = errs.Also(validate.ObjectMetadata(v.GetObjectMeta()).ViaField("metadata"))
+	errs = errs.Also(v.Spec.Validate(ctx))
+	return errs
+}
+
+// Validate VerificationPolicySpec
+func (vs *VerificationPolicySpec) Validate(ctx context.Context) (errs *apis.FieldError) {
+	if len(vs.Resources) == 0 {
+		errs = errs.Also(apis.ErrMissingField("resources"))
+	}
+	if len(vs.Authorities) == 0 {
+		errs = errs.Also(apis.ErrMissingField("authorities"))
+	}
+	for i, a := range vs.Authorities {
+		if a.Key != nil {
+			errs = errs.Also(a.Key.Validate(ctx).ViaFieldIndex("key", i))
+		}
+	}
+	return errs
+}
+
+// Validate KeyRef
+func (key *KeyRef) Validate(ctx context.Context) (errs *apis.FieldError) {
+	if key.Data == "" && key.SecretRef == nil {
+		errs = errs.Also(apis.ErrMissingOneOf("data", "secretref"))
+	}
+
+	if key.Data != "" && key.SecretRef != nil {
+		errs = errs.Also(apis.ErrMultipleOneOf("data", "secretref"))
+	}
+	return errs
```

Comment:
is hashAlgorithm required, or does it have a default?

Reply:
hashAlgorithm is handled in `hashAlgorithm` later. So it is not required here.

Reply:
No, it is called in reconciler code.
Yes that makes sense!

Reply:
I have added the validation into webhook